feat(authncache): add cache layer for public authn keys (WIP)

LeraKravts opened the pull request:
# Which Problems Are Solved

- Introduce a new cache layer for public authentication keys (`authnKeys`) to improve lookup performance.
- Prepare structure for Memory/Redis backends via `connector.StartCache` framework.

# How the Problems Are Solved

- Added a new struct `CachedPublicKey` under `internal/authncache`
- Registered a new `authnKeys` cache in `query/start.go` using existing cache connector
- Defined indexing via `AuthnKeyIndex` (KeyID, InstanceID, UserType)
- Implemented cache invalidation through `projection.AuthNKeyProjection.RegisterCacheInvalidation`

# Additional Changes

- Added `cache.PurposeAuthNKeys` to define a new cache purpose
- Introduced new `AuthnKeyIndex` index values (KeyID, InstanceID, UserType)
- Implemented `CachedPublicKey` model in `internal/authncache`
- Registered `authnKeys` cache in `query/start.go` using `StartCache` connector pattern
- Registered cache invalidation via `projection.AuthNKeyProjection.RegisterCacheInvalidation`

# Additional Context

This is a **draft PR** for early review. It is not ready for merge yet.
- Config integration (`defaults.yaml`) is planned but not yet finalized
- Data write & load logic (e.g. projection handlers) will be handled in follow-up commits
- Related to: https://github.com/zitadel/zitadel/issues/10096

fforootd commented:
Nice PR! Thank you. Let us review this.

CC @muhlemmer since you worked on that issue before

LeraKravts commented:
Just a quick note: there might be a lint issue, similar to what happened in #10097. It looks like the dev container provided by the project is still using Go 1.22, so I might have missed some formatting or linting differences from the current CI setup.

LeraKravts commented:
Hey Silvan, 

Thank you very much for the review. I've addressed all review comments:

-Cache key now includes `instanceID:keyID:identifier`
-Expiration is checked before use; expired entries fallback to DB
-SQL filters (`enabled = true`, `expiration > now`) ensure only valid keys are cached
-Removed unused index. 

Let me know that there's anything else that need to be fixed.

LeraKravts commented:
Hi Silvan @adlerhurst, no problem — thank you for the update!

Just to clarify: should I wait for your team's decision before starting on these tasks?
https://github.com/zitadel/zitadel/issues/10095
https://github.com/zitadel/zitadel/issues/10101

adlerhurst commented:
> Hi Silvan @adlerhurst, no problem — thank you for the update!
> 
> Just to clarify: should I wait for your team's decision before starting on these tasks? #10095 #10101

I gave feedback on the two issues. Thanks a lot for asking.

adlerhurst commented on internal/query/cache.go:
```diff
@@ -30,6 +32,21 @@ func startCaches(background context.Context, connectors connector.Connectors, in
+			AuthnKeyIndexKeyIDAndIdentifier,
+		},
+		cache.PurposeAuthNKeys,
+		connectors.Config.Instance,
```
Please create add an `AuthNKey` field in the `CachesConfig` to allow custom configuration of this cache.

elinashoko commented:
We haven't heard in a while, so closing the PR. Please reopen if needed.
